bokeh/sphinxext/bokeh_example_sampledata.py

```diff
@@ -0,0 +1,149 @@
+
+class sampledata_list(nodes.General, nodes.Element):
+    def __init__(self, *args, **kwargs):
+        self.sampledata_key = kwargs.pop("sampledata_key")
+        super().__init__(*args, **kwargs)
+
+class BokehExampleSampledataDirective(BokehDirective):
+    has_content = False
+    required_arguments = 1
+
+    def run(self):
+        return [sampledata_list('', sampledata_key=self.arguments[0])]
+
+
+def setup(app):
+    """ Required Sphinx extension setup function. """
+    app.add_node(sampledata_list)
+    app.add_directive("bokeh-example-sampledata", BokehExampleSampledataDirective)
+    app.connect('doctree-resolved', process_sampledata_xrefs)
+    app.connect('env-purge-doc', purge_xrefs)
+    app.connect('env-merge-info', merge_xrefs)
+    return PARALLEL_SAFE
+
+# -----------------------------------------------------------------------------
+# Private API
+# -----------------------------------------------------------------------------
+
+def purge_xrefs(app, env, docname):
+    if not hasattr(env, 'all_sampledata_xrefs'):
+        return
+
+    env.all_sampledata_xrefs = [xref for xref in env.all_sampledata_xrefs if xref['docname'] != docname]
+
+def merge_xrefs(app, env, docnames, other):
+    if not hasattr(env, 'all_sampledata_xrefs'):
+        env.all_sampledata_xrefs = []
+
+    if hasattr(other, 'all_sampledata_xrefs'):
+        env.all_sampledata_xrefs.extend(other.all_sampledata_xrefs)
+
+def process_sampledata_xrefs(app, doctree, fromdocname):
+
+    env = app.builder.env
+
+    if not hasattr(env, 'all_sampledata_xrefs'):
+        env.all_sampledata_xrefs = []
+
+    for node in doctree.traverse(sampledata_list):
+        sampladata_key = node.sampledata_key
+        content = []
+
+        this_sampladata_refs = []
+        for sample_info in env.all_sampledata_xrefs:
+            if sample_info['keyword'] == sampladata_key:
+                this_sampladata_refs.append(sample_info)
+
+        _len = len(this_sampladata_refs)
+        para = nodes.paragraph()
+        if _len:
+            s = '' if _len==1 else 's'
+            description = (
+            _(f'Check out the demonstration{s} for the ``{sampladata_key}`` sampledata. See the example{s} '))
+        else:
+            description = (
+            _(f'Well, there is no documented standalone example with the ``{sampladata_key}`` sampledata'))
+        para += nodes.Text(description, description)
+        for i, sample_info in enumerate(this_sampladata_refs):
+            # Create all references
+            newnode = nodes.reference('', '')
+            ref_name = basename(sample_info['docname'])
+            innernode = nodes.emphasis(_(ref_name), _(ref_name))
+            newnode['refdocname'] = sample_info['docname']
+            newnode['refuri'] = app.builder.get_relative_uri(
+                fromdocname, sample_info['docname'])
+            # TODO missing tags for references in user_guide
```

Comment:
The extention trys to link to the [`user_guide`](https://docs.bokeh.org/en/latest/docs/user_guide/data.html#mapping-marker-types), which was unintended but is quite nice. But the correct tag is missing and I don't know how to get this information. Any hint or idea is appreciated.

![iris](https://user-images.githubusercontent.com/68053396/164990097-0fecbcd0-ca2d-4463-85ae-d3f9f8341613.PNG)

Reply:
I don't know offhand either, and in fact I am not sure there is any guarantee to be one. The inline examples have never necessarily had anchors or references attached specifically to them. They definitley don't have "names" that match up to a docname 1-1 in the sense the gallery examples do. I think ideally the output would be a list because then we could always link to the source, and then only link to a live example for the gallery, e.g. something like:
```
* iris.py (gallery, source)
* iris_splom.py (gallery, source)
* foo.py (source)
```
where "gallery" is a link to a gallery doc page and "source" is `bokeh-tree` link to the actual source file in the GH repo. 

All that said I think for now to keep things simple it would be fine to just filter the user guide entries out, and only link to gallery doc pages.

Reply:
Alternatively if we want to do this I think we would have to abandon using the docname and instead use the actual filename for the tag, but I still think finding the right place to link to inside the user guide may be an issue (or maybe I am mistaken)